Improve error reporting when user fails to attach `@MultipartForm` annotation - similar to #6738

## Description

If a user forgets to attach `@MultipartForm` annotation to RESTeasy controller parameter, the error returned is non-descript and not very helpful.

Maybe this could be better, similar to the solution in #6738

```kt
import org.jboss.resteasy.reactive.RestForm
import org.jboss.resteasy.reactive.multipart.FileUpload

class SourceArchiveUploadForm {
    @RestForm
    lateinit var file: FileUpload
}

@Path("/upload")
class UploadsController {

    @POST
    @Path("/script")
    @Produces(MediaType.TEXT_PLAIN)
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    fun uploadScript(upload: SourceArchiveUploadForm): String {
        return UploadsControllerService.handleFileUpload(upload)
    }
}
```

This will fail with:
```
HTTP/1.1 415 Unsupported Media Type
content-length: 0
```

**Full output of test run log:**
<details>

		    Jul 11, 2021 5:09:50 PM org.jboss.threads.Version <clinit>
		    INFO: JBoss Threads version 3.4.0.Final
		    Jul 11, 2021 5:09:52 PM org.jboss.threads.Version <clinit>
		    INFO: JBoss Threads version 3.4.0.Final
		    Jul 11, 2021 5:09:52 PM io.quarkus.bootstrap.runner.Timing printStartupTime
		    INFO: Quarkus 2.0.1.Final on JVM started in 2.630s. Listening on: http://localhost:8081
		    Jul 11, 2021 5:09:52 PM io.quarkus.bootstrap.runner.Timing printStartupTime
		    INFO: Profile test activated. 
		    Jul 11, 2021 5:09:52 PM io.quarkus.bootstrap.runner.Timing printStartupTime
		    INFO: Installed features: [cdi, config-yaml, jaxrs-client-reactive, kotlin, qute, rest-client, rest-client-mutiny, rest-client-reactive, rest-client-reactive-jackson, resteasy-reactive, resteasy-reactive-qute, smallrye-context-propagation, smallrye-health, vertx]
		    HTTP/1.1 415 Unsupported Media Type
		    content-length: 0
		    
		    2 expectations failed.
		    Expected status code <200> but was <415>.
		    
		    Response body doesn't match expectation.
		    Expected: is "ok"
		      Actual: 
		    
		    java.lang.AssertionError: 2 expectations failed.
		    Expected status code <200> but was <415>.
		    
		    Response body doesn't match expectation.
		    Expected: is "ok"
		      Actual: 
		    
			    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
			    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
			    at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
			    at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
			    at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:72)
			    at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:59)
			    at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84)
			    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
			    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
			    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
			    at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:493)
			    at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
			    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
			    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
			    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
			    at io.restassured.internal.ResponseSpecificationImpl.forceValidateResponse(ResponseSpecificationImpl.groovy:660)
			    at io.restassured.module.kotlin.extensions.RestAssuredKotlinExtensionsKt$Then$2.invoke(RestAssuredKotlinExtensions.kt:61)
			    at io.restassured.module.kotlin.extensions.RestAssuredKotlinExtensionsKt$Then$2.invoke(RestAssuredKotlinExtensions.kt:60)
			    at io.restassured.module.kotlin.extensions.RestAssuredKotlinExtensionsKt$doIfValidatableResponseImpl$1.invoke(RestAssuredKotlinExtensions.kt:80)
			    at io.restassured.module.kotlin.extensions.RestAssuredKotlinExtensionsKt$doIfValidatableResponseImpl$1.invoke(RestAssuredKotlinExtensions.kt:78)
			    at io.restassured.module.kotlin.extensions.RestAssuredKotlinExtensionsKt.Then(RestAssuredKotlinExtensions.kt:60)
			    at org.hasura.controllers.UploadsControllerTest.uploadScript(UploadsControllerTest.kt:29)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
			    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
			    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
			    at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:989)
			    at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:875)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
			    at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
			    at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
			    at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
			    at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
			    at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
			    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
			    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
			    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
			    at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
			    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
			    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
			    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
			    at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
			    at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
			    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
			    at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
			    at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
			    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
			    at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
			    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
			    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
			    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
			    at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
			    at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
			    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
			    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
			    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
			    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
			    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
			    at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
			    at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
			    at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
			    at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
			    at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
			    at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
			    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
			    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
			    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
			    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
			    at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
			    at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
			    at com.sun.proxy.$Proxy2.stop(Unknown Source)
			    at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
			    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
			    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
			    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
			    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
			    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
			    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
			    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
			    at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
			    at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
			    at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
			    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
			    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
			    at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
			    at java.base/java.lang.Thread.run(Thread.java:829)
		    
		    
		    WARNING: An illegal reflective access operation has occurred
		    WARNING: Illegal reflective access by org.codehaus.groovy.vmplugin.v9.Java9 (file:/C:/Users/rayga/.gradle/caches/modules-2/files-2.1/org.codehaus.groovy/groovy/3.0.8/43a7ecf3b8a9d7dd9fc8d468ae96b97cff4616df/groovy-3.0.8.jar) to constructor java.lang.AssertionError(java.lang.String)
		    WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.vmplugin.v9.Java9
		    WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
		    WARNING: All illegal access operations will be denied in a future release
		    UploadsControllerTest > uploadScript() FAILED
		        java.lang.AssertionError at UploadsControllerTest.kt:29
		    Jul 11, 2021 5:09:55 PM io.quarkus.bootstrap.runner.Timing printStopTime
		    INFO: Quarkus stopped in 1.371s
		    1 test completed, 1 failed
		    > Task :test FAILED
		    FAILURE: Build failed with an exception.
		    * What went wrong:
		    Execution failed for task ':test'.
		    > There were failing tests. See the report at: file:///C:/Users/rayga/Projects/tmp/polyglot-actions-service/quarkus-kotlin-gradle/build/reports/tests/test/index.html
		    * Try:
		    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
		    * Get more help at https://help.gradle.org
		    BUILD FAILED in 11s
		    11 actionable tasks: 7 executed, 4 up-to-date


</details>

--- a/docs/src/main/asciidoc/resteasy-reactive.adoc
+++ b/docs/src/main/asciidoc/resteasy-reactive.adoc
@@ -466,6 +466,8 @@ public class Endpoint {
 The use of link:{resteasy-reactive-common-api}/org/jboss/resteasy/reactive/MultipartForm.html[`@MultipartForm`] as
 method parameter makes RESTEasy Reactive handle the request as a multipart form request.
 
+TIP: The use of `@MultipartForm` is actually unnecessary as RESTEasy Reactive can infer this information from the use of `@Consumes(MediaType.MULTIPART_FORM_DATA)`
+
 WARNING: When handling file uploads, it is very important to move the file to permanent storage (like a database, a dedicated file system or a cloud storage) in your code that handles the POJO.
 Otherwise, the file will no longer be accessible when the request terminates.
 Moreoever if `quarkus.http.body.delete-uploaded-files-on-end` is set to true, Quarkus will delete the uploaded file when the HTTP response is sent. If the setting is disabled,

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartResource.java
@@ -39,7 +39,7 @@ public class MultipartResource {
     @Produces(MediaType.TEXT_PLAIN)
     @Consumes(MediaType.MULTIPART_FORM_DATA)
     @Path("/blocking")
-    public Response blocking(@DefaultValue("1") @RestQuery Integer times, @MultipartForm FormData formData) throws IOException {
+    public Response blocking(@DefaultValue("1") @RestQuery Integer times, FormData formData) throws IOException {
         if (!BlockingOperationControl.isBlockingAllowed()) {
             throw new RuntimeException("should not have dispatched");
         }
@@ -57,7 +57,7 @@ public class MultipartResource {
     @Produces(MediaType.TEXT_PLAIN)
     @Consumes(MediaType.MULTIPART_FORM_DATA)
     @Path("/same-name")
-    public String sameName(@MultipartForm FormDataSameFileName formData) {
+    public String sameName(FormDataSameFileName formData) {
         if (BlockingOperationControl.isBlockingAllowed()) {
             throw new RuntimeException("should not have dispatched");
         }

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -424,6 +424,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                         existingConverters, additionalReaders,
                         anns, paramType, errorLocation, false, hasRuntimeConverters, pathParameters,
                         currentMethodInfo.parameterName(i),
+                        consumes,
                         methodContext);
                 suspended |= parameterResult.isSuspended();
                 sse |= parameterResult.isSse();
@@ -806,6 +807,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             Map<String, String> existingConverters, AdditionalReaders additionalReaders,
             Map<DotName, AnnotationInstance> anns, Type paramType, String errorLocation, boolean field,
             boolean hasRuntimeConverters, Set<String> pathParameters, String sourceName,
+            String[] declaredConsumes,
             Map<String, Object> methodContext) {
         PARAM builder = createIndexedParam()
                 .setCurrentClassInfo(currentClassInfo)
@@ -933,7 +935,15 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                 if (field) {
                     return builder;
                 }
-                builder.setType(ParameterType.BODY);
+                if ((declaredConsumes != null) && (declaredConsumes.length == 1)
+                        && (MediaType.MULTIPART_FORM_DATA.equals(declaredConsumes[0]))) {
+                    // in this case it is safe to assume that we are consuming multipart data
+                    // we already don't allow multipart to be used along with body in the same method,
+                    // so this is completely safe
+                    builder.setType(ParameterType.MULTI_PART_FORM);
+                } else {
+                    builder.setType(ParameterType.BODY);
+                }
             }
         }
         builder.setSingle(true);

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/ServerEndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/ServerEndpointIndexer.java
@@ -191,7 +191,7 @@ public class ServerEndpointIndexer
                     additionalReaders,
                     annotations, field.type(), field.toString(), true, hasRuntimeConverters,
                     // We don't support annotation-less path params in injectable beans: only annotations
-                    Collections.emptySet(), field.name(), new HashMap<>());
+                    Collections.emptySet(), field.name(), EMPTY_STRING_ARRAY, new HashMap<>());
             if ((result.getType() != null) && (result.getType() != ParameterType.BEAN)) {
                 //BODY means no annotation, so for fields not injectable
                 fieldExtractors.put(field, result);

